src/Catalyst.Core.Lib/FileSystem/FileSystem.cs

```diff
@@ -24,17 +24,104 @@
+using ProtoBuf;
 
 namespace Catalyst.Core.Lib.FileSystem
 {
     // ReSharper disable once ClassWithVirtualMembersNeverInherited.Global
-    public class FileSystem : System.IO.Abstractions.FileSystem, IFileSystem
+    public class FileSystem : System.IO.Abstractions.FileSystem, Catalyst.Abstractions.FileSystem.IFileSystem
     {
+        private static readonly byte[] emptyData = new byte[0];
+
+        /// <summary>
+        ///   Creates a stream that can read the supplied <see cref="Cid"/>.
+        /// </summary>
+        /// <param name="id">
+        ///   The identifier of some content.
+        /// </param>
+        /// <param name="blockService">
+        ///   The source of the cid's data.
+        /// </param>
+        /// <param name="keyChain">
+        ///   Used to decypt the protected data blocks.
+        /// </param>
+        /// <param name="cancel">
+        ///   Is used to stop the task.  When cancelled, the <see cref="TaskCanceledException"/> is raised.
+        /// </param>
+        /// <returns>
+        ///   A task that represents the asynchronous operation. The task's value is
+        ///   a <see cref="Stream"/> that produces the content of the <paramref name="id"/>.
+        /// </returns>
+        /// <remarks>
+        ///  The id's <see cref="Cid.ContentType"/> is used to determine how to read
+        ///  the conent.
+        /// </remarks>
+        public static Task<Stream> CreateReadStreamAsync(Cid id,
+            IBlockApi blockService,
+            KeyChain keyChain,
+            CancellationToken cancel)
+        {
+            // TODO: A content-type registry should be used.
+            if (id.ContentType == "dag-pb")
+                return CreateDagProtoBufStreamAsync(id, blockService, keyChain, cancel);
+            else if (id.ContentType == "raw")
+                return CreateRawStreamAsync(id, blockService, keyChain, cancel);
+            else if (id.ContentType == "cms")
+                return CreateCmsStreamAsync(id, blockService, keyChain, cancel);
+            else
+                throw new NotSupportedException($"Cannot read content type '{id.ContentType}'.");
+        }
+
+        private static async Task<Stream> CreateRawStreamAsync(Cid id,
+            IBlockApi blockService,
+            KeyChain keyChain,
+            CancellationToken cancel)
+        {
+            var block = await blockService.GetAsync(id, cancel).ConfigureAwait(false);
+            return block.DataStream;
+        }
+
+        private static async Task<Stream> CreateDagProtoBufStreamAsync(Cid id,
+            IBlockApi blockService,
+            KeyChain keyChain,
+            CancellationToken cancel)
+        {
+            var block = await blockService.GetAsync(id, cancel).ConfigureAwait(false);
+            var dag = new DagNode(block.DataStream);
+            var dm = Serializer.Deserialize<DataMessage>(dag.DataStream);
+
+            if (dm.Type != DataType.File)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4612924)